### PR TITLE
feat: dask expr `cast`

### PR DESCRIPTION
**Pull request by aidoskanapyanov**
## What type of PR is this? (check all applicable)

- [ ] 💾 Refactor
- [x] ✨ Feature
- [ ] 🐛 Bug Fix
- [ ] 🔧 Optimization
- [ ] 📝 Documentation
- [ ] ✅ Test
- [ ] 🐳 Other

## Related issues 

- Related issue #637 
- Closes #

## Checklist

- [x] Code follows style guide (ruff)
- [ ] Tests added 
- [ ] Documented the changes

## If you have comments or can explain your changes, please do so below.

Implemented without preserving a dtype backend for now. Need to dig more into that :P

### CODE REVIEW

File: narwhals/_dask/utils.py [hunk 1]
```diff
@@ -73,3 +78,47 @@ def parse_exprs_and_named_exprs(
...
+    if isinstance_or_issubclass(dtype, dtypes.UInt32):
+        return "uint32"
+    if isinstance_or_issubclass(dtype, dtypes.UInt16):
+        return "uint16"
+    if isinstance_or_issubclass(dtype, dtypes.UInt8):
+        return "uint8"
+    if isinstance_or_issubclass(dtype, dtypes.String):
+        pd = get_pandas()
+
+        if pd is not None and parse_version(pd.__version__) >= parse_version("2.0.0"):
```
Comment (FBruzzesi) [inline narwhals/_dask/utils.py:109]: Nitpick or trick I learned from Marco:

```suggestion
        if (pd := get_pandas()) is not None and parse_version(pd.__version__) >= parse_version("2.0.0"):
```

File: narwhals/_dask/expr.py [hunk 2]
```diff
@@ -610,6 +611,21 @@ def dt(self: Self) -> DaskExprDateTimeNamespace:
     def name(self: Self) -> DaskExprNameNamespace:
         return DaskExprNameNamespace(self)
 
+    def cast(
+        self,
+        dtype: Any,
+    ) -> Self:
+        def func(_input: Any, dtype: Any) -> Any:
```
Comment (FBruzzesi) [inline narwhals/_dask/expr.py:618]:
```suggestion
    def cast(
        self: Self,
        dtype: DType | type[DType],
    ) -> Self:
        def func(_input: Any, dtype: : DType | type[DType]) -> Any:
```

which would require to add

```py
if TYPE_CHECKING:
    from narwhals.dtypes import DType
```

**Review by FBruzzesi — COMMENTED** [review]
I love this! Thanks for the effort! 🙌🏼
I left a few nitpicks here and there 😇

**Comment by aidoskanapyanov** [comment]
Some CI failures:
- Coverage is dropping for some reason for `pytest-38 (3.8, windows-latest)`
- Also, modin constructor is not failing, but marked as failing in pytest

**Comment by FBruzzesi** [comment]
> * Coverage is dropping for some reason for `pytest-38 (3.8, windows-latest)`

I will take a look and see if I can spot anything

> * Also, modin constructor is not failing, but marked as failing in pytest

That's typically good news even though... why is that happening now 😂?

File: narwhals/_dask/utils.py [hunk 3]
```diff
@@ -73,3 +78,47 @@ def parse_exprs_and_named_exprs(
...
+    if isinstance_or_issubclass(dtype, dtypes.Boolean):
+        return "bool"
+    if isinstance_or_issubclass(dtype, dtypes.Categorical):
+        return "category"
+    if isinstance_or_issubclass(dtype, dtypes.Datetime):
+        return "datetime64[us]"
+    if isinstance_or_issubclass(dtype, dtypes.Duration):
+        return "timedelta64[ns]"
+
+    msg = f"Unknown dtype: {dtype}"
```
Comment (FBruzzesi) [inline narwhals/_dask/utils.py:123]: For coverage, it should be enough to add a `pragma: no cover` here:

```suggestion
    msg = f"Unknown dtype: {dtype}"  # pragma: no cover
```

**Comment by aidoskanapyanov** [comment]
> I wonder why `windows` and `linux` pytest coverage CI show different results. For some reason windows run is flagging lots of lines as uncovered, whereas linux is flagging them as green.

Yeah, nevermind that, I figure CI coverage reports are a bit more involved than I understand rn. Gotta dig into it first.

**Comment by FBruzzesi** [comment]
Hey there! 

I think the problem with coverage for python 3.8 is that dask is entirely skipped (min python version is 3.9), therefore as we increase total number of functions, the coverage will gradually decrease.

I think you just unluckily hit the threshold number to push the total below 90%.

As all other workflows are passing, I would not worry too much about it. I think we will need to be smarter about it or come with another threshold value.

**Review by FBruzzesi — APPROVED** [review]
We adjusted the threshold to 85%, now some other unrelated CI error appeared 😂

> uv pip install --pre --extra-index-url https://pypi.anaconda.org/scientific-python-nightly-wheels/simple pandas --system
  shell: /usr/bin/bash -e {0}
  env:
    pythonLocation: /opt/hostedtoolcache/Python/[3](https://github.com/narwhals-dev/narwhals/actions/runs/10548871513/job/29223261679#step:9:3).12.5/x64
    PKG_CONFIG_PATH: /opt/hostedtoolcache/Python/3.12.5/x6[4](https://github.com/narwhals-dev/narwhals/actions/runs/10548871513/job/29223261679#step:9:4)/lib/pkgconfig
    Python_ROOT_DIR: /opt/hostedtoolcache/Python/3.12.[5](https://github.com/narwhals-dev/narwhals/actions/runs/10548871513/job/29223261679#step:9:5)/x64
    Python2_ROOT_DIR: /opt/hostedtoolcache/Python/3.12.5/x[6](https://github.com/narwhals-dev/narwhals/actions/runs/10548871513/job/29223261679#step:9:6)4
    Python3_ROOT_DIR: /opt/hostedtoolcache/Python/3.12.5/x64
    LD_LIBRARY_PATH: /opt/hostedtoolcache/Python/3.12.5/x64/lib
error: HTTP status server error (503 Service Unavailable) for url (https://pypi.anaconda.org/scientific-python-nightly-wheels/simple/pandas/)
Error: Process completed with exit code 2.

**Comment by aidoskanapyanov** [comment]
Yeah, server error on the anaconda side by the looks of it. I've rerun the CI manually, it's passing now 😄

**Pull request merged**